File: app/src/aether/tools/write_file.py
"""Write File tool — create or overwrite a file."""

from __future__ import annotations

import os

from aether.tools.base import AetherTool, ToolParam, ToolResult


class WriteFileTool(AetherTool):
    name = "write_file"
    description = "Write content to a file. Creates the file if it doesn't exist, overwrites if it does. Creates parent directories as needed."
    status_text = "Writing file..."
    parameters = [
        ToolParam(name="path", type="string", description="File path to write to"),
        ToolParam(name="content", type="string", description="Content to write to the file"),
    ]

    def __init__(self, working_dir: str = "."):
        self.working_dir = os.path.abspath(working_dir)

    def _resolve(self, path: str) -> str:
        resolved = os.path.abspath(os.path.join(self.working_dir, path))
        if not resolved.startswith(self.working_dir):
            raise PermissionError(f"Path escapes working directory: {path}")
        return resolved

    async def execute(self, path: str, content: str) -> ToolResult:
        resolved = self._resolve(path)

        try:
            os.makedirs(os.path.dirname(resolved), exist_ok=True)
            with open(resolved, "w", encoding="utf-8") as f:
                f.write(content)
            size = os.path.getsize(resolved)
            return ToolResult.success(
                f"Written {size:,} bytes to {path}",
                path=path, size=size,
            )
        except PermissionError as e:
            return ToolResult.fail(f"Permission denied: {e}")
        except Exception as e:
            return ToolResult.fail(f"Error writing file: {e}")

File: agent/internal/tools/builtin/write_file.go
package builtin

import (
	"context"
	"os"
	"path/filepath"

	"github.com/suryaumapathy/core-ai/agent/internal/tools"
)

type WriteFileTool struct{}

func (t *WriteFileTool) Definition() tools.Definition {
	return tools.Definition{
		Name:        "write_file",
		Description: "Write text content to a file in the working directory.",
		StatusText:  "Writing file...",
		Parameters: []tools.Param{
			{Name: "path", Type: "string", Description: "Relative file path", Required: true},
			{Name: "content", Type: "string", Description: "File content", Required: true},
			{Name: "append", Type: "boolean", Description: "Append instead of overwrite", Required: false, Default: false},
		},
	}
}

func (t *WriteFileTool) Execute(ctx context.Context, call tools.Call) tools.Result {
	_ = ctx
	path, _ := call.Args["path"].(string)
	content, _ := call.Args["content"].(string)
	appendMode, _ := call.Args["append"].(bool)

	resolved, err := resolveSafePath(call.Ctx.WorkingDir, path)
	if err != nil {
		return tools.Fail("Invalid path: "+err.Error(), nil)
	}
	if err := os.MkdirAll(filepath.Dir(resolved), 0o755); err != nil {
		return tools.Fail("Failed to prepare parent directory: "+err.Error(), nil)
	}

	flags := os.O_CREATE | os.O_WRONLY
	if appendMode {
		flags |= os.O_APPEND
	} else {
		flags |= os.O_TRUNC
	}
	f, err := os.OpenFile(resolved, flags, 0o644)
	if err != nil {
		return tools.Fail("Failed to open file: "+err.Error(), nil)
	}
	defer f.Close()
	if _, err := f.WriteString(content); err != nil {
		return tools.Fail("Failed to write file: "+err.Error(), nil)
	}
	return tools.Success("File written successfully.", map[string]any{"path": resolved, "bytes": len(content), "append": appendMode})
}

var _ tools.Tool = (*WriteFileTool)(nil)
